### ISSUE TITLE
major lag, every 0.5-1s

### ISSUE THREAD

**Issue by sunbeam906** [gamelaster/saboteur-lua-hook#1]
Hi there.

I've tested this on the current GOG version, console works, but the game lags badly. Every 1s it's stuttering/lagging.

Best regards,
Sun

**Comment by sunbeam906** [comment]
Hi again.

If you alt-tab from the game, it will freeze and can't get back in. That's because, I'm assuming here, the device gets lost on alt-tabbing (am playing in full screen, not windowed).

EDIT #1: Yeah, if I run it in windowed mode (same 1920x1080 with "Windowed 1" in config.ini), it lags a bit less and no freezing on alt-tab.

EDIT #2: Mmm, the more I alt-tab, the worse the lagging gets. I will record a video, maybe that helps.

BR,
Sun

**Comment by gamelaster** [comment]
Hi @sunbeam906 ,

I know about this issue, although, no idea why it happens. Although, I have newer lua hook which is better, but it's not in final state, but I might release it soon.

**Comment by sunbeam906** [comment]
> Hi @sunbeam906 ,
> 
> I know about this issue, although, no idea why it happens. Although, I have newer lua hook which is better, but it's not in final state, but I might release it soon.

Sure thing, in the meantime I've had to really understand what the hell "execute" and other names you've used in the source code relate to Lua-wise. I would appreciate it if you use Lua API names for easier discovery in OTHER games that might use Lua 5.1.

I was able to find the 5.1 source code and a build of lua5.1.dll compiled with Visual Studio 8 (the same one used for Saboteur). Attached them below, maybe they're useful to you (including the .pdb symbols).

Download >> [lua51.zip](https://github.com/saboteur-team/saboteur-lua-hook/files/7145051/lua51.zip)

With these combined I deducted that the following in your [lua.hpp](https://github.com/saboteur-team/saboteur-lua-hook/blob/master/lua.hpp) file are as follows:

- 0x004041B0 == _luaL_loadstring_
- 0x00401F90 == _lua_pcall_
- 0x006FAB6B == CALL to _lua_setfield_
- 0x00E4DFAC == inlined _fputs_ API

If you document them like this, then anyone looking to replicate what you've done for The Saboteur can also do it in their target game, granted the game uses Lua 5.1 and d3d9.dll rendering. For example, I wanted to replicate this in MDK 2, but that game uses Lua 3.1 and ddraw.dll, so couldn't adapt it.

Wrote about it here: https://fearlessrevolution.com/viewtopic.php?f=4&t=17377.

Cheers,
Sun

P.S.: I own only the GOG version of the game.

**Comment by gamelaster** [comment]
Hi Sun,

Sorry, I didn't expected at the time of making this that anyone will try to inspire by that 😁 
Although, in the better version it's better described, also it's external DLL which hooks D3D9
differently, and in this version it doesn't lag (also it uses ImGUI for testing)

https://cdn.discordapp.com/attachments/347165529917620225/804357680151789628/2021-01-28_15-28-51.mp4

About your situation, yes, Lua 3.1 is probably way more different, also hooking DirectDraw is different, but it shouldn't be hard to hook it.

Also, nice post about your progress, amazing job!

**Comment by sunbeam906** [comment]
Yeah, Lua 3.1 is much simpler, in the sense that you just need to know the Lua APIs, then send the text as a buffer. No need for encoding or any file handles or any of that crap :) But hey, that hook looks cool even if external ;) I doubt people would want to write stuff to console all the time, thus alt-tab when you need it and that's that. Of course, if you fixed the alt-tab issue when in full screen [?] :)
